python_modules/dagster/dagster/core/execution/run_cancellation_thread.py

```diff
@@ -0,0 +1,42 @@
+import threading
+import time
+from typing import cast
+
+from dagster import check
+from dagster.core.instance import DagsterInstance
+from dagster.core.storage.pipeline_run import PipelineRun, PipelineRunStatus
+from dagster.utils import send_interrupt
+
+INTERVAL_SECONDS = 10
+
+
+def _kill_on_cancel(instance: DagsterInstance, run_id):
+    check.inst_param(instance, "run_id", DagsterInstance)
+    check.str_param(run_id, "run_id")
+
+    while True:
+        time.sleep(INTERVAL_SECONDS)
+        run = cast(
+            PipelineRun,
+            check.inst(
+                instance.get_run_by_id(run_id), PipelineRun, "Run not found for cancellation thread"
```

Comment:
how confident are we that instance.get_run_by_id is thread safe on our storages in general? If we're not, we could pass in an instance ref here and construct a new instance within this thread (much like the daemon creates a new instance for each of its threads)

(The in memory storage does not look thread safe, but that may just be for tests?)

Reply:
👍 sounds smart to create a new instance, though it does double our PG connections right?

Reply:
it does, in theory connection pooling helps with that right?